File: src/Parsing.cpp
#include "Parsing.hpp"
#include "CompileError.hpp"

#include <sstream>
#include <assert.h>

namespace Langite {

    std::unique_ptr<AstFile> ParseFile(std::string_view filepath, std::string_view source) {
        Lexer lexer{ filepath, source };
        std::vector<std::unique_ptr<Ast>> expressions;
        while (lexer.PeekKind() != TokenKind::EndOfFile) {
            AllowMultipleNewlines(lexer);
            expressions.push_back(ParseExpression(lexer));
            ExpectNewline(lexer);
        }
        Token endOfFileToken = lexer.NextToken();
        assert(endOfFileToken.Kind == TokenKind::EndOfFile);
        return std::make_unique<AstFile>(std::move(expressions), endOfFileToken);
    }

    std::unique_ptr<AstBlock> ParseBlock(Lexer& lexer) {
        Token openBraceToken = ExpectToken(lexer, TokenKind::OpenBrace);
        std::vector<std::unique_ptr<Ast>> expressions;
        while (lexer.PeekKind() != TokenKind::CloseBrace) {
            AllowMultipleNewlines(lexer);
            expressions.push_back(ParseExpression(lexer));
            ExpectNewline(lexer);
        }
        Token closeBraceToken = ExpectToken(lexer, TokenKind::CloseBrace);
        return std::make_unique<AstBlock>(openBraceToken, std::move(expressions), closeBraceToken);
    }

    std::unique_ptr<AstDeclaration> ParseDeclaration(Lexer& lexer) {
        Token nameToken           = ExpectToken(lexer, TokenKind::Name);
        Token colonToken          = ExpectToken(lexer, TokenKind::Colon);
        std::unique_ptr<Ast> type = ParseExpression(lexer);
        return std::make_unique<AstDeclaration>(nameToken, colonToken, std::move(type));
    }

    std::unique_ptr<AstIf> ParseIf(Lexer& lexer) {
        Token ifToken                       = ExpectToken(lexer, TokenKind::If);
        std::unique_ptr<Ast> condition      = ParseExpression(lexer);
        std::unique_ptr<AstBlock> thenBlock = ParseBlock(lexer);
        std::optional<Token> elseToken;
        std::optional<std::unique_ptr<Ast>> elseScope;
        if (lexer.PeekKind() == TokenKind::Else) {
            elseToken = lexer.NextToken();
            if (lexer.PeekKind() == TokenKind::If)
                elseScope = ParseIf(lexer);
            else
                elseScope = ParseBlock(lexer);
        }
        return std::make_unique<AstIf>(
            ifToken, std::move(condition), std::move(thenBlock), elseToken, std::move(elseScope));
    }

    std::unique_ptr<Ast> ParseExpression(Lexer& lexer) {
        return ParseBinaryExpression(lexer, 0);
    }

    std::unique_ptr<Ast> ParseLeastExpression(Lexer& lexer) {
        return ParseBinaryExpression(lexer, std::numeric_limits<size_t>::max());
    }

    std::unique_ptr<Ast> ParsePrimaryExpression(Lexer& lexer) {
        switch (lexer.PeekKind()) {
            case TokenKind::OpenParenthesis: {
                Token openParenthesisToken      = lexer.NextToken();
                std::unique_ptr<Ast> expression = ParseExpression(lexer);
                Token closeParenthesisToken     = ExpectToken(lexer, TokenKind::CloseParenthesis);
                return std::make_unique<AstParenthesisedExpression>(
                    openParenthesisToken, std::move(expression), closeParenthesisToken);
            }

            case TokenKind::Const: {
                Token constToken = lexer.NextToken();
                Token nameToken;
                if (lexer.PeekKind() == TokenKind::Wildcard)
                    nameToken = lexer.NextToken();
                else
                    nameToken = ExpectToken(lexer, TokenKind::Name);
                std::optional<Token> openSquareBracketToken;
                std::optional<std::vector<std::unique_ptr<AstDeclaration>>> genericParameters;
                std::optional<Token> closeSquareBracketToken;
                if (lexer.PeekKind() == TokenKind::OpenSquareBracket) {
                    openSquareBracketToken = lexer.NextToken();
                    genericParameters      = std::vector<std::unique_ptr<AstDeclaration>>{};
                    while (lexer.PeekKind() != TokenKind::CloseSquareBracket) {
                        genericParameters->push_back(ParseDeclaration(lexer));
                        ExpectCommaOrNewline(lexer);
                    }
                    closeSquareBracketToken = ExpectToken(lexer, TokenKind::CloseSquareBracket);
                }
                std::optional<Token> colonToken;
                std::optional<std::unique_ptr<Ast>> type;
                if (lexer.PeekKind() == TokenKind::Colon) {
                    colonToken = lexer.NextToken();
                    type       = ParseExpression(lexer);
                }
                Token equalToken           = ExpectToken(lexer, TokenKind::Equal);
                std::unique_ptr<Ast> value = ParseExpression(lexer);
                return std::make_unique<AstConstDeclaration>(constToken,
                                                             nameToken,
                                                             openSquareBracketToken,
                                                             std::move(genericParameters),
                                                             closeSquareBracketToken,
                                                             colonToken,
                                                             std::move(type),
                                                             equalToken,
                                                             std::move(value));
            }

            case TokenKind::Name: {
                Token nameToken = lexer.NextToken();
                if (lexer.PeekKind() == TokenKind::Colon) {
                    Token colonToken          = lexer.NextToken();
                    std::unique_ptr<Ast> type = ParseLeastExpression(lexer);
                    return std::make_unique<AstDeclaration>(nameToken, colonToken, std::move(type));
                } else {
                    return std::make_unique<AstName>(nameToken);
                }
            }

            case TokenKind::Wildcard:
                return std::make_unique<AstWildcard>(lexer.NextToken());

            case TokenKind::Integer:
                return std::make_unique<AstInteger>(lexer.NextToken());

            case TokenKind::Float:
                return std::make_unique<AstFloat>(lexer.NextToken());

            case TokenKind::String:
                return std::make_unique<AstString>(lexer.NextToken());

            case TokenKind::Func: {
                Token funcToken            = lexer.NextToken();
                Token openParenthesisToken = ExpectToken(lexer, TokenKind::OpenParenthesis);
                std::vector<std::unique_ptr<AstDeclaration>> parameters;
                while (lexer.PeekKind() != TokenKind::CloseParenthesis) {
                    parameters.push_back(ParseDeclaration(lexer));
                    ExpectCommaOrNewline(lexer);
                }
                Token closeParenthesisToken     = ExpectToken(lexer, TokenKind::CloseParenthesis);
                Token colonToken                = ExpectToken(lexer, TokenKind::Colon);
                std::unique_ptr<Ast> returnType = ParseLeastExpression(lexer);
                std::optional<std::unique_ptr<AstBlock>> body;
                if (lexer.PeekKind() == TokenKind::OpenBrace)
                    body = ParseBlock(lexer);
                return std::make_unique<AstFunction>(funcToken,
                                                     openParenthesisToken,
                                                     std::move(parameters),
                                                     closeParenthesisToken,
                                                     colonToken,
                                                     std::move(returnType),
                                                     std::move(body));
            }

            case TokenKind::Proc: {
                Token procToken            = lexer.NextToken();
                Token openParenthesisToken = ExpectToken(lexer, TokenKind::OpenParenthesis);
                std::vector<std::unique_ptr<AstDeclaration>> parameters;
                while (lexer.PeekKind() != TokenKind::CloseParenthesis) {
                    parameters.push_back(ParseDeclaration(lexer));
                    ExpectCommaOrNewline(lexer);
                }
                Token closeParenthesisToken     = ExpectToken(lexer, TokenKind::CloseParenthesis);
                Token colonToken                = ExpectToken(lexer, TokenKind::Colon);
                std::unique_ptr<Ast> returnType = ParseLeastExpression(lexer);
                std::optional<std::unique_ptr<AstBlock>> body;
                if (lexer.PeekKind() == TokenKind::OpenBrace)
                    body = ParseBlock(lexer);
                return std::make_unique<AstProcedure>(procToken,
                                                      openParenthesisToken,
                                                      std::move(parameters),
                                                      closeParenthesisToken,
                                                      colonToken,
                                                      std::move(returnType),
                                                      std::move(body));
            }

            case TokenKind::Return: {
                Token returnToken = lexer.NextToken();
                std::optional<std::unique_ptr<Ast>> value;
                if (lexer.PeekKind() != TokenKind::EndOfFile &&
                    lexer.PeekKind() != TokenKind::CloseParenthesis &&
                    lexer.PeekKind() != TokenKind::CloseSquareBracket &&
                    lexer.PeekKind() != TokenKind::Newline)
                    value = ParseExpression(lexer);
                return std::make_unique<AstReturn>(returnToken, std::move(value));
            }

            case TokenKind::If:
                return ParseIf(lexer);

            case TokenKind::OpenBrace:
                return ParseBlock(lexer);

            default: {
                Token token = lexer.NextToken();
                throw CompileError{
                    .Location = token.Location,
                    .Message  = (std::stringstream{} << "Expected an expression, but got '"
                                                    << TokenKind_ToString(token.Kind) << '\'')
                                   .str(),
                };
            }
        }
    }

    std::unique_ptr<Ast> ParseBinaryExpression(Lexer& lexer, size_t parentPrecedence) {
        std::unique_ptr<Ast> left; // this is to delay construction of the std::variant

        size_t unaryPrecedence = GetUnaryPrecedence(lexer.PeekKind());
        if (unaryPrecedence > 0) {
            Token operatorToken          = lexer.NextToken();
            std::unique_ptr<Ast> operand = ParseBinaryExpression(lexer, unaryPrecedence);

            left = std::make_unique<AstUnary>(operatorToken, std::move(operand));
        } else {
            left = ParsePrimaryExpression(lexer);
        }

        while (true) {
            switch (lexer.PeekKind()) {
                case TokenKind::Period: {
                    Token periodToken    = lexer.NextToken();
                    Token fieldNameToken = ExpectToken(lexer, TokenKind::Name);

                    left = std::make_unique<AstFieldAccess>(
                        std::move(left), periodToken, fieldNameToken);
                } break;

                case TokenKind::At: {
                    Token atToken                = lexer.NextToken();
                    std::unique_ptr<Ast> indexer = ParseLeastExpression(lexer);

                    left = std::make_unique<AstIndex>(std::move(left), atToken, std::move(indexer));
                } break;

                case TokenKind::OpenParenthesis: {
                    Token openParenthesisToken = lexer.NextToken();
                    std::vector<std::unique_ptr<Ast>> arguments;
                    while (lexer.PeekKind() != TokenKind::CloseParenthesis) {
                        arguments.push_back(ParseExpression(lexer));
                        ExpectCommaOrNewline(lexer);
                    }
                    Token closeParenthesisToken = ExpectToken(lexer, TokenKind::CloseParenthesis);

                    left = std::make_unique<AstCall>(std::move(left),
                                                     openParenthesisToken,
                                                     std::move(arguments),
                                                     closeParenthesisToken);
                } break;

                case TokenKind::OpenSquareBracket: {
                    Token openSquareBracketToken = lexer.NextToken();
                    std::vector<std::unique_ptr<Ast>> genericArguments;
                    while (lexer.PeekKind() != TokenKind::CloseSquareBracket) {
                        genericArguments.push_back(ParseExpression(lexer));
                        ExpectCommaOrNewline(lexer);
                    }
                    Token closeSquareBracketToken =
                        ExpectToken(lexer, TokenKind::CloseSquareBracket);

                    left = std::make_unique<AstGenericInstantiation>(std::move(left),
                                                                     openSquareBracketToken,
                                                                     std::move(genericArguments),
                                                                     closeSquareBracketToken);
                } break;

                default: {
                    size_t binaryPrecedence = GetBinaryPrecedence(lexer.PeekKind());
                    if (binaryPrecedence <= parentPrecedence)
                        goto End;

                    Token operatorToken        = lexer.NextToken();
                    std::unique_ptr<Ast> right = ParseBinaryExpression(lexer, binaryPrecedence);

                    left = std::make_unique<AstBinary>(
                        std::move(left), operatorToken, std::move(right));
                } break;
            }
        }
End:

        return left;
    }

    size_t GetUnaryPrecedence(TokenKind kind) {
        switch (kind) {
            case TokenKind::Plus:
            case TokenKind::Minus:
            case TokenKind::ExclamationMark:
                return 5;

            default:
                return 0;
        }
    }

    size_t GetBinaryPrecedence(TokenKind kind) {
        switch (kind) {
            case TokenKind::Asterisk:
            case TokenKind::Slash:
            case TokenKind::Modulus:
                return 4;

            case TokenKind::Plus:
            case TokenKind::Minus:
                return 3;

            case TokenKind::EqualEqual:
            case TokenKind::ExclamationMarkEqual:
                return 2;

            case TokenKind::LeftArrow:
            case TokenKind::RightArrow:
                return 1;

            default:
                return 0;
        }
    }

    void AllowNewline(Lexer& lexer) {
        if (lexer.PeekKind() == TokenKind::Newline)
            lexer.NextToken();
    }

    void AllowMultipleNewlines(Lexer& lexer) {
        while (lexer.PeekKind() == TokenKind::Newline)
            lexer.NextToken();
    }

    void ExpectNewline(Lexer& lexer) {
        TokenKind kind = lexer.PeekKind();
        if (kind != TokenKind::EndOfFile && kind != TokenKind::CloseParenthesis &&
            kind != TokenKind::CloseBrace && kind != TokenKind::CloseSquareBracket)
            ExpectToken(lexer, TokenKind::Newline);
    }

    void ExpectCommaOrNewline(Lexer& lexer) {
        TokenKind kind = lexer.PeekKind();
        if (kind != TokenKind::EndOfFile && kind != TokenKind::CloseParenthesis &&
            kind != TokenKind::CloseSquareBracket && kind != TokenKind::Newline)
            ExpectToken(lexer, TokenKind::Comma);
        AllowNewline(lexer);
    }

    Token ExpectToken(Lexer& lexer, TokenKind kind) {
        Token token = lexer.NextToken();
        if (token.Kind == kind)
            return token;
        throw CompileError{
            .Location = token.Location,
            .Message =
                (std::stringstream{} << "Expected '" << TokenKind_ToString(kind) << "', but got '"
                                     << TokenKind_ToString(token.Kind) << '\'')
                    .str(),
        };
    }
}

File: src/syntax/parsing.rs
use derive_more::Display;

use super::{
    Ast, AstBody, AstParameter, BinaryOperator, GetLocation, Lexer, LexerError, SourceSpan, Token,
    TokenKind, UnaryOperator,
};

#[derive(Debug, Display)]
pub enum ParsingError<'filepath, 'source> {
    #[display(fmt = "{_0}")]
    LexerError(LexerError<'filepath>),
    #[display(fmt = "{}: Unexpected {token} token", "token.get_location()")]
    UnexpectedToken { token: Token<'filepath, 'source> },
    #[display(fmt = "{}: Unexpected name token, but got {got}", "got.get_location()")]
    ExpectedNameToken { got: Token<'filepath, 'source> },
    #[display(fmt = "{}: Expected {expected} but got {got}", "got.get_location()")]
    ExpectedTokenButGot {
        expected: TokenKind<'source>,
        got: Token<'filepath, 'source>,
    },
    #[display(fmt = "Unexpected end of source file '{filepath}'")]
    UnexpectedEndOfFile { filepath: &'filepath str },
    #[display(fmt = "{location}: Procedure with name must have a body")]
    ProcedureWithNameMustHaveBody { location: SourceSpan<'filepath> },
}

impl<'filepath, 'source> From<LexerError<'filepath>> for ParsingError<'filepath, 'source> {
    fn from(error: LexerError<'filepath>) -> Self {
        ParsingError::LexerError(error)
    }
}

pub fn parse_file<'filepath, 'source>(
    filepath: &'filepath str,
    source: &'source str,
) -> Result<Vec<Ast<'filepath, 'source>>, ParsingError<'filepath, 'source>> {
    let mut lexer = Lexer::new(filepath, source);
    let mut expressions = vec![];
    while lexer.peek_token()?.is_some() {
        while let Some(Token {
            kind: TokenKind::Newline,
            ..
        }) = lexer.peek_token()?
        {
            next_token(&mut lexer)?;
        }
        expressions.push(parse_expression(&mut lexer)?);
        expect_token(&mut lexer, TokenKind::Newline)?;
    }
    Ok(expressions)
}

fn parse_expression<'filepath, 'source>(
    lexer: &mut Lexer<'filepath, 'source>,
) -> Result<Ast<'filepath, 'source>, ParsingError<'filepath, 'source>> {
    parse_binary_expression(lexer, 0)
}

fn parse_primary_expression<'filepath, 'source>(
    lexer: &mut Lexer<'filepath, 'source>,
) -> Result<Ast<'filepath, 'source>, ParsingError<'filepath, 'source>> {
    let mut expression = match next_token(lexer)? {
        name_token @ Token {
            kind: TokenKind::Name(_),
            ..
        } => Ast::Name { name_token },
        integer_token @ Token {
            kind: TokenKind::Integer(_),
            ..
        } => Ast::Integer { integer_token },
        string_token @ Token {
            kind: TokenKind::String(_),
            ..
        } => Ast::String { string_token },
        let_token @ Token {
            kind: TokenKind::LetKeyword,
            ..
        } => {
            _ = let_token;
            todo!();
        }
        var_token @ Token {
            kind: TokenKind::VarKeyword,
            ..
        } => {
            _ = var_token;
            todo!();
        }
        const_token @ Token {
            kind: TokenKind::ConstKeyword,
            ..
        } => {
            _ = const_token;
            todo!();
        }
        proc_token @ Token {
            kind: TokenKind::ProcKeyword,
            ..
        } => {
            let name_token = if let Some(Token {
                kind: TokenKind::Name(_),
                ..
            }) = lexer.peek_token()?
            {
                Some(next_token(lexer)?)
            } else {
                None
            };
            let open_parenthesis_token = expect_token(lexer, TokenKind::OpenParenthesis)?;
            let mut parameters = vec![];
            while !matches!(
                lexer.peek_token()?,
                Some(Token {
                    kind: TokenKind::CloseParenthesis,
                    ..
                })
            ) {
                while let Some(Token {
                    kind: TokenKind::Newline,
                    ..
                }) = lexer.peek_token()?
                {
                    next_token(lexer)?;
                }

                let name_token = next_token(lexer)?;
                if !matches!(name_token.kind, TokenKind::Name(_)) {
                    return Err(ParsingError::ExpectedNameToken { got: name_token });
                }
                let colon_token = expect_token(lexer, TokenKind::Colon)?;
                let typ = parse_expression(lexer)?;
                parameters.push(AstParameter {
                    name_token,
                    colon_token,
                    typ,
                });

                if matches!(
                    lexer.peek_token()?,
                    Some(Token {
                        kind: TokenKind::CloseParenthesis,
                        ..
                    })
                ) {
                    break;
                }
                expect_token(lexer, TokenKind::Comma)?;
                match_token(lexer, &TokenKind::Newline)?;
            }
            let close_parenthesis_token = expect_token(lexer, TokenKind::CloseParenthesis)?;
            let (fat_right_arrow_token, return_type) =
                if let Some(fat_right_arrow) = match_token(lexer, &TokenKind::FatRightArrow)? {
                    let return_type = parse_expression(lexer)?;
                    (Some(fat_right_arrow), Some(return_type))
                } else {
                    (None, None)
                };
            let body = if let Some(do_token) = match_token(lexer, &TokenKind::DoKeyword)? {
                let expression = parse_expression(lexer)?;
                Some(AstBody::DoExpression {
                    do_token,
                    expression: Box::new(expression),
                })
            } else if let Some(Token {
                kind: TokenKind::OpenBrace,
                ..
            }) = lexer.peek_token()?
            {
                Some(AstBody::Block(Box::new(parse_block(lexer)?)))
            } else {
                if name_token.is_some() {
                    return Err(ParsingError::ProcedureWithNameMustHaveBody {
                        location: proc_token.get_location(),
                    });
                }
                None
            };
            Ast::Procedure {
                proc_token,
                name_token,
                open_parenthesis_token,
                parameters,
                close_parenthesis_token,
                fat_right_arrow_token,
                return_type: return_type.map(|return_type| Box::new(return_type)),
                body,
            }
        }
        return_token @ Token {
            kind: TokenKind::ReturnKeyword,
            ..
        } => {
            let value = if !matches!(
                lexer.peek_token()?,
                Some(Token {
                    kind: TokenKind::Newline, // TODO: what if you have `{ return }`, should there be an extra rule here?
                    ..
                })
            ) {
                Some(parse_expression(lexer)?)
            } else {
                None
            };
            Ast::Return {
                return_token,
                value: value.map(|value| Box::new(value)),
            }
        }
        token => return Err(ParsingError::UnexpectedToken { token }),
    };

    // call
    while matches!(
        lexer.peek_token()?,
        Some(Token {
            kind: TokenKind::OpenParenthesis,
            ..
        })
    ) {
        let open_parenthesis_token = expect_token(lexer, TokenKind::OpenParenthesis)?;
        let mut arguments = vec![];
        while !matches!(
            lexer.peek_token()?,
            Some(Token {
                kind: TokenKind::CloseParenthesis,
                ..
            })
        ) {
            while let Some(Token {
                kind: TokenKind::Newline,
                ..
            }) = lexer.peek_token()?
            {
                next_token(lexer)?;
            }

            arguments.push(parse_expression(lexer)?);

            if matches!(
                lexer.peek_token()?,
                Some(Token {
                    kind: TokenKind::CloseParenthesis,
                    ..
                })
            ) {
                break;
            }
            expect_token(lexer, TokenKind::Comma)?;
            match_token(lexer, &TokenKind::Newline)?;
        }
        let close_parenthesis_token = expect_token(lexer, TokenKind::CloseParenthesis)?;
        expression = Ast::Call {
            operand: Box::new(expression),
            open_parenthesis_token,
            arguments,
            close_parenthesis_token,
        };
    }

    Ok(expression)
}

fn parse_binary_expression<'filepath, 'source>(
    lexer: &mut Lexer<'filepath, 'source>,
    parent_precedence: usize,
) -> Result<Ast<'filepath, 'source>, ParsingError<'filepath, 'source>> {
    let mut left = {
        let operator = match lexer.peek_token()? {
            Some(
                plus_token @ Token {
                    kind: TokenKind::Plus,
                    ..
                },
            ) => Some((3, UnaryOperator::Identity { plus_token })),
            Some(
                minus_token @ Token {
                    kind: TokenKind::Minus,
                    ..
                },
            ) => Some((3, UnaryOperator::Negate { minus_token })),
            _ => None,
        };
        if let Some((unary_precedence, operator)) = operator {
            lexer.next_token()?;
            let operand = parse_binary_expression(lexer, unary_precedence)?;
            Ast::Unary {
                operator,
                operand: Box::new(operand),
            }
        } else {
            parse_primary_expression(lexer)?
        }
    };

    loop {
        let (binary_precedence, operator) = {
            match lexer.peek_token()? {
                Some(
                    plus_token @ Token {
                        kind: TokenKind::Plus,
                        ..
                    },
                ) => (1, BinaryOperator::Add { plus_token }),
                Some(
                    minus_token @ Token {
                        kind: TokenKind::Minus,
                        ..
                    },
                ) => (1, BinaryOperator::Subtract { minus_token }),
                Some(
                    asterisk_token @ Token {
                        kind: TokenKind::Asterisk,
                        ..
                    },
                ) => (2, BinaryOperator::Multiply { asterisk_token }),
                Some(
                    slash_token @ Token {
                        kind: TokenKind::Slash,
                        ..
                    },
                ) => (2, BinaryOperator::Divide { slash_token }),
                _ => break,
            }
        };
        if binary_precedence <= parent_precedence {
            break;
        }

        next_token(lexer)?;
        let right = parse_binary_expression(lexer, binary_precedence)?;
        left = Ast::Binary {
            left: Box::new(left),
            operator,
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_block<'filepath, 'source>(
    lexer: &mut Lexer<'filepath, 'source>,
) -> Result<Ast<'filepath, 'source>, ParsingError<'filepath, 'source>> {
    let open_brace_token = expect_token(lexer, TokenKind::OpenBrace)?;
    let mut expressions = vec![];
    while !matches!(
        lexer.peek_token()?,
        Some(Token {
            kind: TokenKind::CloseBrace,
            ..
        })
    ) {
        while let Some(Token {
            kind: TokenKind::Newline,
            ..
        }) = lexer.peek_token()?
        {
            next_token(lexer)?;
        }

        expressions.push(parse_expression(lexer)?);

        expect_token(lexer, TokenKind::Newline)?;
    }
    let close_brace_token = expect_token(lexer, TokenKind::CloseBrace)?;
    Ok(Ast::Block {
        open_brace_token,
        expressions,
        close_brace_token,
    })
}

fn expect_token<'filepath, 'source>(
    lexer: &mut Lexer<'filepath, 'source>,
    expected: TokenKind<'source>,
) -> Result<Token<'filepath, 'source>, ParsingError<'filepath, 'source>> {
    let token = next_token(lexer)?;
    if token.kind != expected {
        return Err(ParsingError::ExpectedTokenButGot {
            expected,
            got: token,
        });
    }
    Ok(token)
}

fn match_token<'filepath, 'source>(
    lexer: &mut Lexer<'filepath, 'source>,
    expected: &TokenKind<'source>,
) -> Result<Option<Token<'filepath, 'source>>, ParsingError<'filepath, 'source>> {
    Ok(
        if let Some(_token) = lexer.peek_token()?.filter(|token| token.kind == *expected) {
            lexer.next_token()?
        } else {
            None
        },
    )
}

fn next_token<'filepath, 'source>(
    lexer: &mut Lexer<'filepath, 'source>,
) -> Result<Token<'filepath, 'source>, ParsingError<'filepath, 'source>> {
    let token = lexer.next_token()?;
    token.ok_or(ParsingError::UnexpectedEndOfFile {
        filepath: lexer.get_filepath(),
    })
}
